cmake/public/cuda.cmake

```diff
@@ -364,20 +364,10 @@ endif()
 
 # Debug and Release symbol support
 if (MSVC)
-  if ((${CMAKE_BUILD_TYPE} MATCHES "Release") OR (${CMAKE_BUILD_TYPE} MATCHES "RelWithDebInfo") OR (${CMAKE_BUILD_TYPE} MATCHES "MinSizeRel"))
-    if (${CAFFE2_USE_MSVC_STATIC_RUNTIME})
-      list(APPEND CUDA_NVCC_FLAGS "-Xcompiler" "-MT")
-    else()
-      list(APPEND CUDA_NVCC_FLAGS "-Xcompiler" "-MD")
-    endif()
-  elseif(${CMAKE_BUILD_TYPE} MATCHES "Debug")
-    if (${CAFFE2_USE_MSVC_STATIC_RUNTIME})
-      list(APPEND CUDA_NVCC_FLAGS "-Xcompiler" "-MTd")
-    else()
-      list(APPEND CUDA_NVCC_FLAGS "-Xcompiler" "-MDd")
-    endif()
+  if (${CAFFE2_USE_MSVC_STATIC_RUNTIME})
+    list(APPEND CUDA_NVCC_FLAGS "-Xcompiler" "-MT$<$<CONFIG:Debug>:d>")
```

Comment:
I'm worried about this usage because it relies on the upstream to support cmake generator expressions. What about changing it to the following style:
```cmake
if (${CAFFE2_USE_MSVC_STATIC_RUNTIME})
    list(APPEND CUDA_NVCC_FLAGS_DEBUG "-Xcompiler" "-MTd")
    list(APPEND CUDA_NVCC_FLAGS_MINSIZEREL "-Xcompiler" "-MT")
    list(APPEND CUDA_NVCC_FLAGS_RELEASE "-Xcompiler" "-MT")
    list(APPEND CUDA_NVCC_FLAGS_RELWITHDEB "-Xcompiler" "-MT")
else()
    list(APPEND CUDA_NVCC_FLAGS_DEBUG "-Xcompiler" "-MDd")
    list(APPEND CUDA_NVCC_FLAGS_MINSIZEREL "-Xcompiler" "-MD")
    list(APPEND CUDA_NVCC_FLAGS_RELEASE "-Xcompiler" "-MD")
    list(APPEND CUDA_NVCC_FLAGS_RELWITHDEB "-Xcompiler" "-MD")
endif()
```
It is safer and the usage of these variables are well documented.

Reply:
Why is it safer? It just looks like a mess.
Documentation exists for generators too and almost all projects seem to use it. Why would upstream not support generators?

Reply:
>  Generators always work.

Not always. Just like in your first commit, it can be used with `set_target_properties` and `LINK_FLAGS`, right?

But since it gets supported in this PR, let's believe it's well tested.
https://gitlab.kitware.com/cmake/cmake/commit/6b190b5c5221d2447bdf19c20243469025d2614e

Reply:
Yeah, sorry, I mean they always work when they are evaluated. As long as they are used in the right place.
I really just need it fixed on the nvcc flags so the config imports but thought it was bad leaving the other usages there.